### ISSUE TITLE
Lobe - Model works perfectly with identical images on Desktop, but not on iOS bootstrap app

### ISSUE THREAD

**Issue by technoplato** [lobe/iOS-bootstrap#22]
Hello, love the app, first of all. This is how training should be!

I did run into a snag: I've successfully trained the model to recognize screenshots from audible, youtube, and apple podcasts and that model works very well on desktop using the "Play" section.

However, when I export the model and use it in the iOS Bootstrap, it does very bad at recognizing any screenshot accurately. 

All I've done with code is replaced `LobeModel` with `SavedModel` directly from Lobe.

Any suggestions? 

Video demo link: https://www.youtube.com/watch?v=_rtKnlmsxzM

**Comment by technoplato** [comment]
Disregard. 

There is a popup when you go to export your model in the desktop application that asks if you'd like to optimize your model.

I clicked optimize the second go round and everything works flawlessly now. Thanks!

**Comment by technoplato** [comment]
Sorry Lobe people for the back and forth, I’m going to reopen this as the performance on device, even with the optimized model is much less accurate than that on the desktop application. 

Any tips would be appreciated.

**Comment by ellbosch** [comment]
Thanks @technoplato—we will be deploying changes as soon as this week to fix these issues. You are correct to note that optimizing your model will not resolve this issue :)

**Comment by technoplato** [comment]
Well that's great to hear! Glad I'm not crazy.

As a side note, is Object Detection getting close or is that still pretty far out in the timeline? 

Thanks!

**Comment by ellbosch** [comment]
Object detection is a rather large release and a top priority, but we can't say more than that. We are starting an insiders ring to test early features with object detection, if you are interested in joining email [lobeai@microsoft.com](mailto:lobeai@microsoft.com).

**Comment by technoplato** [comment]
Totally understand. Quick question: are all models expected to have poor
performance or just the CreateML?

**Comment by ellbosch** [comment]
For the particular bug you posted, it's a bug specific to image processing. There's actually nothing wrong with the Core ML model—it should behave as you would expect. We're just plugging in a bad input to the model which creates the weird results you're seeing.

**Comment by technoplato** [comment]
Now I’m a little more confused. Is the bug with the starter code then?
Could you elaborate on what you mean about image processing?

**Comment by ellbosch** [comment]
You are correct that it's a bug with the starter code. Here's a brief explanation on how iOS-bootstrap works:
1. Video capture: first, the iOS device's camera transmits video feed.
2. Image processing: a frame of the video feed needs formatting (with cropping and orientation adjustments) before we can predict.
3. Prediction: the processed image is assigned a prediction from the Core ML model's classifier.

Step 2 above is what's broken, and should be fixed this week.

I highly recommend [our docs](https://docs.lobe.ai/docs/improving/improving) if you'd like to learn more! Please let me know if you have any other questions.

**Comment by technoplato** [comment]
Haven’t had a chance to dig too deep into the code yet but just wanted to
make sure that (2) is expected even when an image is selected from the
image library.

**Comment by ellbosch** [comment]
Good question! Yes, we also format images selected from device library before sending to prediction (i.e. we need images formatted into a square aspect ratio).

**Comment by technoplato** [comment]
Perfect. So images chosen from the library will suffer the same bug. Will
be looking forward to the update patiently. Thank you guys for your support
here and for creating such an accessible and seamless tool for such a
powerful technology!

**Comment by ellbosch** [comment]
Your welcome :) the PR is here if you'd like a sneak peak of the fix: #23

**Comment by technoplato** [comment]
Super clean PR - love the refactoring and more information.

Unfortunately, it actually looks like the same issue of inaccurate detection (of images that work on Desktop) as `develop` :(

Can I help provide any information for validating / resolving the issue?

**Comment by ellbosch** [comment]
Shoot 😩 I'm sorry this PR didn't fix things. My guess is I was focusing too much on fixing image processing bugs for the camera feed rather than image preview mode, although I know both modes are more reliable with this upcoming PR.

I'll try to repro your specific project. I *think* I know why this might still be breaking on your end. I'll reach back out to you via this thread when I have an update here.

**Comment by technoplato** [comment]
Please don’t apologize!! It’s an awesome PR and I’m getting everything for
free!

I made a gif replicating the issue. I can also share my mode if it helps
just ping me :)

**Comment by ellbosch** [comment]
Hello @technoplato! I believe PR #25 might have resolved this bug. Would you mind validating this by testing the `u/elbosc/63/model-preprocessing` branch with your project?

**Comment by technoplato** [comment]
I can’t wait. I saw release .9 and am going to try and rerun my mode
tomorrow :)

**Comment by technoplato** [comment]
Hey @ellbosch I'm running into 'Network Disconnected' errors when trying to download the application either through https://lobeprod.azureedge.net/downloads/macos/Lobe.zip or through the home page (which I assume uses the same endpoint) 

Can you send me the zipped application directly at halfjew22@gmail.com? Really looking forward to playing with the new tool and demos

**Comment by mbeissinger** [comment]
@technoplato sent you an email with a link as well

**Comment by technoplato** [comment]
Unfortunately, no dice here either. I'm going to upload my training set and see if I'm just doing something wrong: https://drive.google.com/drive/folders/1nGTCSE9dsgz1ZJDVgc4rMGmBoyNJ-wmk?usp=sharing

I'd upload the model as well but I'm suspicious that's what I'm messing up. 

Received the email @mbeissinger. Going to try exporting a new model, but after trying a v.8 model with the new iOS bootstrap that @ellbosch linked above, I'm still getting poor results

**Comment by mbeissinger** [comment]
@technoplato is the model in iOS giving different results than from Lobe's Use tab?

One thing from quickly looking at the dataset -- it is probably overfitting since most of the images in each class look very similar to each other. I would try to prune out images that pretty much look the same and try to capture more varied examples.

**Comment by technoplato** [comment]
@mbeissinger Thank you for the suggestion.

I've got 99% accuracy within the Lobe Desktop application. The issue arises when I attempt to use the model in the ios-bootstrap. I believe Lobe Desktop's almost perfect accuracy implies that, despite a relatively repetitive dataset, the model "functions", even if just in the Desktop application

**Comment by mbeissinger** [comment]
@technoplato ah ok! So is the CoreML model not loading/giving any results or are they different results than in Lobe?

**Comment by technoplato** [comment]
Correct. The CoreML model is giving results but they’re all incorrect and
different than Lobe.

**Comment by mbeissinger** [comment]
@technoplato ok great I'll see if we can replicate

**Comment by technoplato** [comment]
Awesome let me know! I’ll be available to try other things as well to help
troubleshoot if you’d like / need.

Hopefully operator error here. I’m driving Elliott nuts 😜

**Comment by ellbosch** [comment]
@technoplato the only thing driving me nuts is that this bug still exists. Sorry we still haven't resolved this!

**Comment by mbeissinger** [comment]
@technoplato @ellbosch I verified we have automatic tests for the model outputs going through the converters, so likely a bug in the ios processing. I'm working on adding a file picker to the web-bootstrap so it isn't just the webcam -- you should be able to test the same image within Lobe and in the starter project soon. Can you also try it with the Lobe Connect local API using the same image as something you drag into the Use tab?

**Comment by technoplato** [comment]
Sorry for my delay @mbeissinger finally got around to this and the `Lobe Connect Run` mode works perfectly!

```sh
python3 lobeconnect.py
correct label: youtube.png
predicted label:	youtube
confidence:		1
correct label: audible.png
predicted label:	audible
confidence:		1
```

Confirming the React App works as well. Thanks for the updates to allow static image selection!
